### PR TITLE
feat: Implement CSRF protection for public TRPC endpoints

### PR DESCRIPTION
**Pull request by zomars**
## What does this PR do?

Fixes # (issue)

## Requirement/Documentation

- If there is a requirement document, please, share it here.
- If there is ab UI/UX design document, please, share it here.

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] Chore (refactoring code, technical debt, workflow improvements)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

## How should this be tested?

- Are there environment variables that should be set?
- What are the minimal test data to have?
- What is expected (happy path) to have (input and output)?
- Any other important info that could help to test that PR

## Mandatory Tasks

- [ ] Make sure you have self-reviewed the code. A decent size PR without self-review might be rejected.

## Checklist

- I haven't read the [contributing guide](https://github.com/calcom/cal.com/blob/main/CONTRIBUTING.md)
- My code doesn't follow the style guidelines of this project
- I haven't commented my code, particularly in hard-to-understand areas
- I haven't checked if my PR needs changes to the documentation
- I haven't checked if my changes generate no new warnings
- I haven't added tests that prove my fix is effective or that my feature works
- I haven't checked if new and existing unit tests pass locally with my changes

### CODE REVIEW

**Comment by zomars** [comment]
Current dependencies on/for this PR:
* main
  * **PR #12392** <a href="https://app.graphite.dev/github/pr/calcom/cal.com/12392?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/calcom/cal.com/12392?utm_source=stack-comment).

**Comment by ThyMinimalDev** [comment]
Would you mind creating a Loom video to showcase the expected behavior after this pull request? This will be helpful for maintaining context when we later explore handling multiple payment apps. Thank you!

File: apps/web/pages/api/auth/csrf.ts [hunk 1]
```diff
@@ -0,0 +1,25 @@
...
+export default function handler(req: NextApiRequest, res: NextApiResponse) {
+  console.log("✨ Getting CSRF token...");
+  if (req.method === "GET") {
+    const token = randomBytes(28).toString("hex");
+    res.setHeader(
+      "Set-Cookie",
+      serialize("csrf_token", token, {
+        httpOnly: false, // important for reading cookie on the client
+        maxAge: undefined, // expire with session
+        sameSite: "strict",
```
Comment (hariombalhara) [inline apps/web/pages/api/auth/csrf.ts:14]: Note that in embed we are in third party context and this cookie would not be sent from within the embed iframe potentially breaking embeds

**Review by hariombalhara — CHANGES_REQUESTED** [review]
We need to confirm that embed won't break with this csrf cookie requirement

File: apps/web/pages/api/auth/csrf.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
+import { serialize } from "cookie";
+import { randomBytes } from "crypto";
+import type { NextApiRequest, NextApiResponse } from "next";
+
+export default function handler(req: NextApiRequest, res: NextApiResponse) {
```
[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/web/pages/api/auth/csrf.ts:5]:
```ts
export const setCsrfToken = (res: ServerResponse) => {
  const token = randomBytes(28).toString("hex");
  res.setHeader(
    "Set-Cookie",
    serialize("csrf_token", token, {
      httpOnly: false, // important for reading cookie on the client
      maxAge: undefined, // expire with session
      sameSite: "strict",
      path: "/",
      secure: process.env.NODE_ENV === "production",
    })
  );
};

export default function handler(req: NextApiRequest, res: NextApiResponse) {
  console.log("✨ Getting CSRF token...");
  if (req.method === "GET") {
    setCsrfToken(res);
    res.status(200).json({ message: "OK!" });
    res.end();
  } else {
    // Handle any other HTTP method
    return res.status(501);
  }
}
```
and then in the getServerSideprops of public pages

```ts
export const getServerSideProps: GetServerSideProps<UserPageProps> = async (context) => {
  setCsrfToken(context.res);
  
 }
```

File: apps/web/pages/api/auth/csrf.ts (hunk 1 above)
Comment (zomars) [reply apps/web/pages/api/auth/csrf.ts:14]: This is a good point!

**Comment by PeerRich** [comment]
lets throw an error for people who are using it:

"You are using a private API endpoint which is for internal use only. To access the public API, please reach out to engineering@cal.com"

(do we have engineering@cal.com ?)

File: apps/web/pages/api/auth/csrf.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/web/pages/api/auth/csrf.ts:5]: I think wealso need to find a way to not have the ssr trpc client throw an error

[the remark this answers is not in the document]
Comment (ThyMinimalDev) [reply apps/web/pages/api/auth/csrf.ts:5]: not 100% sure about this. but I think we need to have a `Secret`, use this secret in `ssrInit` to set a header for the `ssr trpc client`, if this header containing the secret is present, skip the `CSRF middleware` when `ssr trpc client` fetches data.

[the remark this answers is not in the document]
Comment (zomars) [reply apps/web/pages/api/auth/csrf.ts:5]: I'm looking if we can re-use the [csrfToken that next-auth generates](https://next-auth.js.org/getting-started/client#getcsrftoken)

**Comment by aecorredor** [comment]
Could https://github.com/calcom/cal.com/pull/12491 be merged in + deployed before getting the CSRF fix in?

We've been working around not having and endpoint for getting an event type by slug by leveraging the public TRPC endpoints.

File: apps/web/pages/api/auth/csrf.ts
Comment (zomars) [inline apps/web/pages/api/auth/csrf.ts:1]: Need to rename this endpoints as it is conflicting with the next-auth one.

**Comment by keithwillcode** [comment]
Closing due to staleness.

**Pull request closed without merging**